### PR TITLE
[Clang] Refactor uses of  `Cand->Function` in SemaOverload.cpp

### PR DESCRIPTION
**Pull request by vortex73**
- [ ] adds checks to called functions containing `Cand->Function` as an argument.
- [ ] Assigned `Cand->Function` as a `FunctionDecl*` to enhance readablity.
Solves: #98769 and #98942

### CODE REVIEW

**Comment by vortex73** [comment]
@haberman Please review.

File: clang/lib/Sema/SemaOverload.cpp [hunk 1]
```diff
@@ -11846,19 +11854,21 @@ static void DiagnoseFailedEnableIfAttr(Sema &S, OverloadCandidate *Cand) {
 }
 
 static void DiagnoseFailedExplicitSpec(Sema &S, OverloadCandidate *Cand) {
-  ExplicitSpecifier ES = ExplicitSpecifier::getFromDecl(Cand->Function);
+  assert(Cand->Function && "Candidate must be a function");
+  Function *CandFunc = Cand->Function;
```
Comment (shafik) [inline clang/lib/Sema/SemaOverload.cpp:11858]: We use `Fn` elsewhere, we should be consistent throughout the file.

File: clang/lib/Sema/SemaOverload.cpp [hunk 2]
```diff
@@ -9994,8 +9995,9 @@ Sema::AddArgumentDependentLookupCandidates(DeclarationName Name,
                                    CandEnd = CandidateSet.end();
        Cand != CandEnd; ++Cand)
     if (Cand->Function) {
-      Fns.erase(Cand->Function);
-      if (FunctionTemplateDecl *FunTmpl = Cand->Function->getPrimaryTemplate())
+      FunctionDecl *CandFunc = Cand->Function;
```
Comment (shafik) [inline clang/lib/Sema/SemaOverload.cpp:9998]: We use `Fn` elsewhere, we should be consistent with the rest of the file.

**Review by shafik — APPROVED** [review]
Thank you for this refactor, just a couple of minor comments but otherwise looks good. Please fix first and then merge after it passes tests, if you need someone to merge for you please let me know.

File: clang/lib/Sema/SemaOverload.cpp (hunk 1 above)
Comment (shafik) [inline clang/lib/Sema/SemaOverload.cpp:11858]:
```suggestion
  FunctionDecl *CandFunc = Cand->Function;
```
Whoops missed this the first time around, looks like a typo, please fix.

**Review by shafik — CHANGES_REQUESTED** [review]

**Comment by vortex73** [comment]
@shafik I've made the changes. Please merge as I do not have write access.

**Comment by vortex73** [comment]
@shafik are any changes required for this PR?

**Review by shafik — APPROVED** [review]
LGTM

**Pull request merged**